src/index/LocatedTriples.h

```diff
@@ -0,0 +1,174 @@
+// Copyright 2023 - 2024, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Authors:
+//    2023 Hannah Bast <bast@cs.uni-freiburg.de>
+//    2024 Julian Mundhahs <mundhahj@tf.uni-freiburg.de>
+
+#pragma once
+
+#include "engine/idTable/IdTable.h"
+#include "global/IdTriple.h"
+#include "index/CompressedRelation.h"
+#include "util/HashMap.h"
+
+class Permutation;
+
+IdTriple permute(const IdTriple& triple, const std::array<size_t, 3>& keyOrder);
+
+// A triple and its block in a particular permutation.
+// For a detailed definition of all border cases, see the definition at
+// the end of this file.
+struct LocatedTriple {
+  // The index of the block, according to the definition above.
+  size_t blockIndex_;
+  // The `Id`s of the triple in the order of the permutation. For example,
+  // for an object pertaining to the OPS permutation: `id1` is the object,
+  // `id2` is the predicate, and `id3` is the subject.
+  IdTriple triple_;
+
+  // Flag that is true if the given triple is inserted and false if it
+  // is deleted.
+  bool shouldTripleExist_;
+
+  // Locate the given triples in the given permutation.
+  static std::vector<LocatedTriple> locateTriplesInPermutation(
+      const std::vector<IdTriple>& triples, const Permutation& permutation,
+      bool shouldExist);
+
+  bool operator==(const LocatedTriple&) const = default;
+
+  friend std::ostream& operator<<(std::ostream& os, const LocatedTriple& lt) {
+    os << "LT(" << lt.blockIndex_ << " " << lt.triple_[0] << " "
+       << lt.triple_[1] << " " << lt.triple_[2] << " " << lt.shouldTripleExist_
+       << ")";
+    return os;
+  }
+};
+
+// A sorted set of located triples. In `LocatedTriplesPerBlock` below, we use
+// this to store all located triples with the same `blockIndex_`.
+//
+// NOTE: We could also overload `std::less` here, but the explicit specification
+// of the order makes it clearer.
+struct LocatedTripleCompare {
+  bool operator()(const LocatedTriple& x, const LocatedTriple& y) const {
+    return x.triple_ < y.triple_;
```

Comment:
Currently, this shouldn't happen. If a triple is inserted and then deleted it is removed from the inserts beforehand. Though it is probably a good to add it the added/inserted triples even if it was removed from the opposite triple right before. E.g. adding an existing triple and then removing it could otherwise lead the triple to exist, although it was deleted.

Reply:
The book keeping of which triples were already inserted/deleted happens in `DeltaTriples`. `LocatedTriplesPerBlock::add` asserts that something was actually inserted, i.e. no duplicate w.r.t. to the compare was inserted. This includes both actual duplicates and insert/delete pairs.